consumer/consumererror/README.md

```diff
@@ -0,0 +1,124 @@
+
+**Communicating network error codes**: Errors should allow embedding information
+necessary for the Collector to act as a proxy for a backend, i.e. relay a status
+code returned from a backend in a response to a system upstream from the
+Collector.
+
+## Creating an error
+
+To create a new error, call the `consumererror.NewError` function with the error
+you want to wrap and any options that add additional metadata. Errors are either
+considered permanent or retryable, depending on whether the
+`WithRetryable[Signal]` option is passed. A permanent error indicates that the
+error will always occur for a given set of telemetry items and should be
+dropped. The permanence of an error can be checked with
+`consumererror.IsPermanent`.
+
+To help communicate information about the error to the caller, options with
+additional metadata may be included. While some of the options are semantically
+mutually-exclusive and shouldn't be combined, any set of options can be used
+together and the package will determine which option takes precedence.
+
+### WithRejectedCount(rejected int)
+
+Include a count of the items that were permanently rejected (spans,
+datapoints, log records, etc.). The caller should have a full count of rejected
+items, so this option is only needed to indicate a partial success.
+
+When using this option in an exporter or other component dealing with mapping
+non-pdata formats, the rejected count should be based on the count of pdata
+items that failed.
+
+### WithRetryable\[Signal\](pdata, ...RetryOption)
+
+Indicate that a temporary condition is the cause of an error and that the
+request should be retried with the default delay. Use of this option means that
+an error is not considered permanent.
+
+#### WithRetryDelay(delay time.Duration)
+
+Indicate that the payload should be retried after a certain amount of time.
+
+### WithHTTPStatus(int)
+
+Annotate the error with an HTTP status code obtained from a response during
+exporting.
+
+### WithGRPCStatus(*status.Status)
+
+Annotate the error with a gRPC status code obtained from a response during
+exporting.
+
+## Reading from an error
+
+The `consumererror.Error` type supports the following methods. Each method has a
+method signature like `(data, bool)` with the second value indicating whether
+the option was passed during the error's creation.
+
+- `Count`: Get the count of rejected items.
+- `Retryable[Signal]`: Gets the information necessary to retry the request for a
+  given signal.
+- `ToHTTP`: Returns an integer representing the HTTP status code. If both
+  `WithHTTPStatus` and `WithGRPCStatus` were passed, the HTTP status is used
+  first and gRPC status second.
+- `ToGRPC`: Returns a `*status.Status` object representing the status returned
+  from the gRPC server. If both `WithHTTPStatus` and `WithGRPCStatus` were
+  passed, the gRPC status is used first and HTTP status second.
+
+## Other considerations
+
+### Asynchronous processing
```

Comment:
Are we able to account for non-linear data flow in some way? For example, when a receiver is shared by multiple pipelines, an error may be returned from any of these pipelines, but I believe we currently do not have a generalized ability to retry only to the pipeline which returned the error. We may need such a mechanism within the consumer package (or fanoutconsumer) in order to avoid sending duplicate data in response to errors.

Reply:
There are two fanout locations that may want to utilize these errors to do retries:
* [fanoutconsumer](https://github.com/open-telemetry/opentelemetry-collector/blob/edf7aca8522a8396ec155eabf8c1119690ed6e2d/internal/fanoutconsumer/traces.go#L47)
* [exporterhelper](https://github.com/open-telemetry/opentelemetry-collector/blob/cc88aee67530478255ecd4f4f5930390c88f5cb1/exporter/exporterhelper/retry_sender.go#L25)

I've tried my best to document the possible situations we may encounter and advice on how to handle them. If I've missed or misrepresented anything, let me know.